source/dub/internal/utils.d

```diff
@@ -342,6 +342,54 @@ string getDUBVersion()
 	return verstr;
 }
 
+/// Returns: current executable's path if running as dub executable,
+/// or as dub library, look for dub executable in the following locations
+///   - current working directory
+///   - same directory as compilerBinary (if supplied)
+///   - all components of the $PATH variable
+/// Throws: an Exception if nothing succeded.
+public string getDUBExePath(in string compilerBinary=null)
+{
+	version(DubApplication) {
+		import std.file : thisExePath;
+		return thisExePath();
+	}
+	else {
+		// this must be dub as a library
+		import std.file : exists, getcwd, isFile;
+		import std.path : buildPath, pathSplitter;
+		import std.process : environment;
+
+		version(Windows) {
+			enum exeName = "dub.exe";
+		}
+		else {
+			enum exeName = "dub";
+		}
+
+		const cwdDub = buildPath(getcwd(), exeName);
+		if (exists(cwdDub) && isFile(cwdDub)) {
+			return cwdDub;
+		}
+
+		if (compilerBinary) {
+			const exe = buildPath(dirName(compilerBinary), exeName);
+			if (exists(exe) && isFile(exe)) {
+				return exe;
+			}
+		}
```

Comment:
Have a look at https://github.com/dlang/dmd/blob/74281dc320329953b44a20f4aa2dfaacc967cc64/src/dmd/frontend.d#L121 for an idea on how this could be simplified with ranges